Overhaul/rewrite CONTRIBUTING.md (#2019)
OpenTabletDriver/OpenTabletDriver: CONTRIBUTING.md

Thank you for showing interest in the development of OpenTabletDriver!

Our goal with this document is to describe how contributions to OpenTabletDriver should be made,
depending on what kind of change is being performed.

> 

Reviewer: ```suggestion

With any contribution you make, we ask that your commits are named sensibly and similarly to the ones in our [history](https://github.com/OpenTabletDriver/OpenTabletDriver/commits/master).
At the very least, sentence-case your commits, and do consider prefixing the commit message with a component name in similar fashion to `Updater: Add file IncludeList`.
If the reasoning for a commit is potentially cryptic, consider adding a commit description to it, succinctly explaining why it was needed.

```

# Tablet Configuration Contributions

> Tablet configurations are the core at what defines a tablet within OpenTabletDriver. It provides

Author: Dislike this wording. How about instead:

```suggestion
Tablet configurations define tablets OpenTabletDriver can detect and operate. They provide
```

> Tablet configurations are the core at what defines a tablet within OpenTabletDriver. It provides

Reviewer: LGTM

  | ---------------- | ------------------------------------------------------------------------------------------------------------------------- |
  | Supported        | All tablet functions are fully supported without issue.                                                                   |
  | Missing Features | Some tablet functions are unsupported (for now).                                    |
  | Has Quirks       | Some tablet functions work, but with undesired behavior or requiring workarounds to get functioning.                      |
  | Broken           | Tested but does not work. This is here for historical purposes, *no new configuration will be accepted with this status.* |
  | Untested         | Entirely untested, potentially ported from other drivers or from documentation. We will potentially accept these, but strongly prefer verification being performed. |

- Include sources to help with verification of the tablet's specifications. This could be the
  manufacturer's specification sheet of the tablet, the HID report descriptor, or an educated guess
  by taking `MaxX` and `MaxY` and converting it into millimeter units. Please note that specs published

>   by manufacturers have unfortunately been known to be wrong on occasion - avoid trusting them
>   without verifying.

Reviewer 2: ```suggestion
  by manufacturers have unfortunately been known to be wrong on occasion - do not trust them
  without verifying them first.
```

  respect `.editorconfig` will do this automatically.
- Split up individual bug fixes or feature additions into separate pull requests. If a pull request
  depends on another, you can make the dependency be merged first by putting `- [ ] depends on
  #<PR_ID>` in the description. This makes it easier to pinpoint why changes were made along with being
  faster to review, and easier to revert if necessary.
- Try to avoid opening PRs with many commits - squash commits where reasonable. If you're having
  trouble combining commits, it may be a sign to make multiple PRs instead. However, do not
  over-squash, preserving history and keeping commits readable is important too.
- If the pull request closes an issue, link that issue with [closing
  keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
  referring to that issue.

## Setting up your environment

OpenTabletDriver is a fairly simple project to set up for development. As it's a .NET project, you
ideally want an IDE capable of working with C# and .NET projects. VSCode, Rider, Visual Studio, etc

> are all suitable. Note that VSCode will require you to install C# extensions first.

Reviewer 2: May be worth stating that most use JetBrains Rider, along with linking each of the listed IDEs.

To get started, create a fork of the repository on GitHub. Then, clone the fork you made, and ensure
you're capable of building OpenTabletDriver, by running either `build.ps1` (for Windows), or

> `build.sh` (for Linux). If successful, these scripts will produce executables in a newly created

Reviewer 2: ```suggestion
`build.sh` (for Linux and macOS). If successful, these scripts will produce executables in a newly created
```

- Use spaces for indentation, do not use the tab character. We use 4 space indentation. IDEs that
  respect `.editorconfig` will do this automatically.
- Split up individual bug fixes or feature additions into separate pull requests. If a pull request
  depends on another, you can make the dependency be merged first by putting `- [ ] depends on
  #<PR_ID>` in the description. This makes it easier to pinpoint why changes were made along with being
  faster to review, and easier to revert if necessary.
- Try to avoid opening PRs with many commits - squash commits where reasonable. If you're having
  trouble combining commits, it may be a sign to make multiple PRs instead. However, do not
  over-squash, preserving history and keeping commits readable is important too.
- If the pull request closes an issue, link that issue with [closing
  keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
  referring to that issue.

## Setting up your environment

> OpenTabletDriver is a fairly simple project to set up for development. As it's a .NET project, you
> ideally want an IDE capable of working with C# and .NET projects.
> [VSCode](https://code.visualstudio.com/) (will require you to install C# extensions first),
> [Rider](https://www.jetbrains.com/rider/) (Used by many c# devs and is our recommended IDE), [Visual
> Studio](https://visualstudio.microsoft.com/), etc are all suitable. Note that OpenTabletDriver being
> an Open Source project means a permanent [free license to use
> Rider](https://www.jetbrains.com/community/opensource/#support) can be obtained.

Reviewer: ```suggestion
OpenTabletDriver is a fairly simple project to set up for development. As it's a .NET project, you
ideally want an IDE capable of working with C# and .NET projects, such as:

- [VSCode](https://code.visualstudio.com/) (with C# extensions installed)
- [Rider](https://www.jetbrains.com/rider/) (used by many C# developers and is our recommended IDE)
- [Visual Studio](https://visualstudio.microsoft.com/)

Note that OpenTabletDriver being an open source project means [a permanent free license to use
Rider](https://www.jetbrains.com/community/opensource/#support) can be obtained.
```
Paragraph was a bit of a mess with all the parentheses and links inline.
Also a few touches, like capturing the entire noun with adjectives when linking the Rider license.

Reviewer 2: > permanent free license to use

It must be updated yearly, and only free to use for open source projects.

ideally want an IDE capable of working with C# and .NET projects. VSCode, Rider, Visual Studio, etc
are all suitable.

To get started, create a fork of the repository on GitHub. Then, clone the fork you made, and ensure
you're capable of building OpenTabletDriver, by running either `build.ps1` (for Windows), or
`build.sh` (for Linux). If successful, these scripts will produce executables in a newly created
`bin` subdirectory. You can then run OpenTabletDriver by running the executable for the Daemon and
the UX. For all intents and purposes, this will be a fully functional, though potentially unstable
(as it was compiled from `master`) version of OpenTabletDriver that you can use normally, though do
note there are cases where plugins may not be updated to work on `master`-built versions.

Once you've confirmed you are able to build the application on your computer, you can begin making
changes. Before working, however, ensure you create a new branch to work in - we
recommend/soft-require that pull requests never be made on your fork's `master` branch - this makes
it harder to make further PRs, as you'll have to wait for your original to be merged before you can

> make another (as your new PR will contain commits from your previous).

Author: osu!'s repo recommends similar, as well as "git best practices" that I've read. Your fork's master should pretty much always stay in line with upstream's master.

> make another (as your new PR will contain commits from your previous).

Reviewer: Fair enough, safe to call this one resolved then.

Once you're done, push the created branch to your fork, and open the PR. Fill out the title and
description of what the PR accomplishes, remembering to link any fixed issues. If any part of our
continuous integration fails, fix it - **ideally by amending the most recent commit and force
pushing,** rather than creating commits purely to fix CI, small spelling mistakes, etc.

If asked to make changes or fixes, please make them through GitHub's **batch commit** feature or
locally, and avoid making a commit per fix - making a single "apply fixes from code review" commit
is preferred. If you disagree with a suggested change, provide your reasoning by replying to the
review comment. Resolve applied fixes, if they are not resolved by GitHub automatically.

## Unit-Testing OpenTabletDriver

> If you'd like to test OpenTabletDriver locally (our CI runs our unit tests on every PR
> automatically), you can run our unit testing through `dotnet` - simply running:
> 
> ```
> dotnet test
> ```
> 
> In the project root will run all tests and report status. The unit tests for the most part concern
> the daemon part of the driver, which performs the bulk of the "work" and thus is critical to test.

Reviewer: ```suggestion
Our CI runs unit tests on each new commit pushed on a PR, but if you'd like to test OpenTabletDriver locally, you can do so by running the command `dotnet test OpenTabletDriver` from the project root, which will run all tests and report status.
The unit tests for the most part concern the daemon part of the driver, which performs the bulk of the "work" and thus is critical to test.
```
Rewording. Also specified the project of the test command, in case it is needed as is the case with dotnet format.
